test/common/event/dispatcher_impl_test.cc

```diff
@@ -325,7 +325,7 @@ class TimerImplTimingTest : public testing::Test {
     while (true) {
       dispatcher.run(Dispatcher::RunType::NonBlock);
       if (timer.enabled()) {
-        time_system.sleep(std::chrono::microseconds(1));
```

Comment:
Isn't the purpose of this test to advance the time and wait until callbacks are run, and then see how long it took?  Why would this not replace the run / advanceTime combo with sleep() which guarantees callbacks?

Reply:
Single-threaded unit tests need to run advanceTime, which puts callbacks into a runnable state but does not run them.

This is because the running of callbacks doesn't happen from TImeSystem; it happens from the dispatcher, so single-threaded tests run the loop after advancing time.

One thing I'd be open to if you prefer is to remove sleep() (rather than changing its meaning) and have instead:
  * advanceTimeBlockForCallbacks()
  * advanceTimeAsync()

which would be exactly the same behavior as sleep() and advanceTime() respectively, but would require all call-sites to change to the appropriate method. WDYT?

I was thinking about this because Nighthawk tests currently call sleep() and we would need to change them to advanceTime() I think, and it might be better to have them fail to build than to deadlock during tests. @oschaaf FYI

Reply:
> which would be exactly the same behavior as sleep() and advanceTime() respectively, but would require all call-sites to change to the appropriate method. WDYT?

+1 from me. I think this is less surprising.

Reply:
(It could even be a single API with a single enum parap saying block/async/etc.)

Reply:
I like the single-API-with-mode in principle and I started to make that change, and the enum is a little awkward at call-sites: `Event::TestTimeSystem::Mode::WaitForTimers`

So I'm leaning toward two APIs.

Reply:
OK thanks for the explanation, nbd either way.